packages/flutter/test/widgets/semantics_traversal_test.dart

```diff
@@ -0,0 +1,365 @@
+      'C': const Offset(40.0, 0.0) & tenByTen,
+      'D': const Offset(60.0, 0.0) & tenByTen,
+      'E': const Offset(0.0, 20.0) & tenByTen,
+      'F': const Offset(0.0, 40.0) & tenByTen,
+      'G': const Offset(20.0, 20.0) & (tenByTen * 2.0),
+      'H': const Offset(60.0, 20.0) & tenByTen,
+      'I': const Offset(60.0, 40.0) & tenByTen,
+      'J': const Offset(0.0, 60.0) & tenByTen,
+      'K': const Offset(20.0, 60.0) & tenByTen,
+      'L': const Offset(40.0, 60.0) & tenByTen,
+      'M': const Offset(60.0, 60.0) & tenByTen,
+    };
+
+    await tester.test(
+      textDirection: TextDirection.ltr,
+      children: children,
+      expectedTraversal: 'A B C D E F G H I J K L M',
+    );
+
+    await tester.test(
+      textDirection: TextDirection.rtl,
+      children: children,
+      expectedTraversal: 'D C B A H I G E F M L K J',
+    );
+  });
+
+  // The following test tests traversal of the simplest "knot", which is two
+  // nodes overlapping both vertically and horizontally. For example:
+  //
+  // ┌─────────┐
+  // │         │
+  // │   A     │
+  // │     ┌───┼─────┐
+  // │     │   │     │
+  // └─────┼───┘     │
+  //       │     B   │
+  //       │         │
+  //       └─────────┘
+  //
+  // The outcome depends on the relative positions of the centers of `Rect`s of
+  // their respective boxes, specifically the direction (i.e. angle) of the
+  // vector pointing from A to B. We test different angles, one for each octant:
+  //
+  //  -3π/4 -π/2  -π/4
+  //      ╲   │   ╱
+  //       ╲ 1│2 ╱
+  //        ╲ │ ╱
+  //     i=0 ╲│╱ 3
+  //  π ──────┼────── 0
+  //       7 ╱│╲ 4
+  //        ╱ │ ╲
+  //       ╱ 6│5 ╲
+  //      ╱   │   ╲
+  //   3π/4  π/2   π/4
+  //
+  // For LTR, angles falling into octants 3, 4, 5, and 6, produce A -> B, all
+  // others produce B -> A.
+  //
+  // For RTL, angles falling into octants 5, 6, 7, and 0, produce A -> B, all
+  // others produce B -> A.
+  testTraversal('Semantics sorts knots', (TraversalTester tester) async {
+    const double start = -math.pi + math.pi / 8.0;
+
+    for (int i = 0; i < 8; i += 1) {
+      final double angle = start + i.toDouble() * math.pi / 4.0;
+      final double dx = math.cos(angle) * 5.0;
+      final double dy = math.sin(angle) * 5.0;
+
+      final Map<String, Rect> children = <String, Rect>{
+        'A': const Offset(10.0, 10.0) & tenByTen,
+        'B': new Offset(10.0 + dx, 10.0 + dy) & tenByTen,
+      };
+
+      try {
+        await tester.test(
+          textDirection: TextDirection.ltr,
+          children: children,
+          expectedTraversal: 3 <= i && i <= 6 ? 'A B' : 'B A',
+        );
+
+        await tester.test(
+          textDirection: TextDirection.rtl,
+          children: children,
+          expectedTraversal: 1 <= i && i <= 4 ? 'B A' : 'A B',
+        );
+      } catch (error) {
+        fail(
+          'Test failed with i == $i, angle == ${angle / math.pi}π\n'
+          '$error'
+        );
+      }
+    }
+  });
```

Comment:
Love all these test cases. Should we add some test cases that use artificial sort keys?

Reply:
We have them. They just live in one of the `semantics_#_test.dart` files. We should revise this at some point. It's hard to tell what each of those files is testing.